internal/pkg/storageos/volume_labels.go

```diff
@@ -162,3 +180,77 @@ func (c *Client) EnsureReplicas(ctx context.Context, key client.ObjectKey, desir
 	}
 	return observeErr(nil)
 }
+
+// EnsureFailureMode ensures that the desired failure mode has been applied to
+// the StorageOS volume.
+func (c *Client) EnsureFailureMode(ctx context.Context, key client.ObjectKey, desired string) error {
+	funcName := "ensure_failure_mode"
+	start := time.Now()
+	defer func() {
+		metrics.Latency.Observe(funcName, time.Since(start))
+	}()
+	observeErr := func(e error) error {
+		metrics.Errors.Increment(funcName, e)
+		return e
+	}
+
+	ctx = c.AddToken(ctx)
+
+	vol, err := c.getVolume(ctx, key)
+	if err != nil {
+		return observeErr(err)
+	}
+
+	var current string
+
+	// If label already set, get value.
+	for k, v := range vol.Labels {
+		if k == ReservedLabelFailureMode {
+			current = v
+			break
+		}
+	}
+
+	// No change required.
+	if current == desired {
+		return nil
+	}
+
+	// Parse the label value to determine the failure mode intent or threshold.
+	mode, threshold, err := ParseFailureMode(desired)
+	if err != nil {
+		return observeErr(err)
+	}
+
+	// Apply update.
+	if _, resp, err := c.api.SetFailureMode(ctx, vol.NamespaceID, vol.Id, api.SetFailureModeRequest{Mode: mode, FailureThreshold: threshold, Version: vol.Version}, nil); err != nil {
+		return observeErr(api.MapAPIError(err, resp))
+	}
+	return observeErr(nil)
+}
+
+// ParseFailureMode parses a string and returns either a failure mode intent or
+// a threshold, if set.  Only one of the intent or threshold should be set.
+func ParseFailureMode(s string) (api.FailureModeIntent, uint64, error) {
+	//  Return defaults if no value set.
+	if s == "" {
```

Comment:
This expression should move int `switch` `case "", FailureModeHard`

Reply:
yes, that's better.